apps/balancer-tools/src/app/poolsimulator/(components)/PoolParamsForm.tsx

```diff
@@ -55,110 +58,155 @@ const inputMapper: InputMapperType = {
     {
       name: "beta",
       label: "Beta",
       placeholder: "Enter beta",
       unit: "",
+      transformFromDataToForm: (n) => n,
+      transformFromFormToData: (n) => n,
     },
     {
       name: "lambda",
       label: "Lambda",
       placeholder: "Enter lambda",
       unit: "",
+      transformFromDataToForm: (n) => n,
+      transformFromFormToData: (n) => n,
     },
     {
       name: "c",
       label: "C",
       placeholder: "Enter c",
       unit: "",
+      transformFromDataToForm: (n) => n,
+      transformFromFormToData: (n) => n,
     },
     {
       name: "s",
       label: "S",
       placeholder: "Enter s",
       unit: "",
+      transformFromDataToForm: (n) => n,
+      transformFromFormToData: (n) => n,
     },
   ],
 };
 
-export function PoolParamsForm() {
-  const { push } = useRouter();
-  const {
-    setIsGraphLoading,
-    setInitialData,
-    setCustomData,
-    initialData,
-    poolType,
-  } = usePoolSimulator();
+export function PoolParamsForm({
+  extraOnSubmit,
+}: {
+  extraOnSubmit?: (data: AnalysisData, tabClicked: boolean) => void;
+}) {
+  const { data, setData, isCustomData } = usePoolFormContext();
 
   const form = useForm({
-    resolver: zodResolver(schemaMapper[poolType]),
-    mode: "onSubmit",
+    resolver: zodResolver(schemaMapper[data.poolType]),
+    mode: "onChange",
   });
   const {
     register,
     setValue,
-    getValues,
     clearErrors,
+    watch,
     formState: { errors },
   } = form;
 
-  const onSubmit = (data: FieldValues) => {
-    setIsGraphLoading(true);
+  const getOnSubmit = (tabClicked: boolean) => (fieldData: FieldValues) => {
+    const hasNullData = inputMapper[data.poolType].reduce(
+      (sum, input) => sum || !fieldData[input.name],
+      false
+    );
+    if (Object.keys(errors).length || hasNullData) return;
     const dataWithPoolType = {
       poolParams: Object.fromEntries(
-        inputMapper[poolType].map((input) => [input.name, data[input.name]])
+        inputMapper[data.poolType].map((input) => [
+          input.name,
+          input.transformFromFormToData(fieldData[input.name]),
+        ])
       ),
-      tokens: initialData.tokens,
-      poolType,
+      tokens: fieldData.tokens,
+      poolType: data.poolType,
     };
-
-    setInitialData(dataWithPoolType as AnalysisData);
-    setCustomData(dataWithPoolType as AnalysisData);
-    push("/poolsimulator/analysis");
+    setData(dataWithPoolType as AnalysisData);
+    extraOnSubmit?.(dataWithPoolType as AnalysisData, tabClicked);
   };
 
+  function checkDataIsEqualForm() {
+    const fieldData = watch();
+    const dataTransformed = {
+      poolParams: Object.fromEntries(
+        inputMapper[data.poolType].map((input) => [
+          input.name,
+          input.transformFromDataToForm(fieldData[input.name]),
+        ])
+      ),
+      tokens: fieldData.tokens,
+      poolType: data.poolType,
+    };
+    return JSON.stringify(dataTransformed) == JSON.stringify(data);
```

Comment:
I don't think JSON.stringify guarantees the order of keys is the same, but I think in your scenario you only want to ensure that the two objects have the same data, right?